fwutil/lib.py

```diff
@@ -0,0 +1,987 @@
+                module_component_map[module_name][component_name] = component
+                #component_firmware_version = component.get_firmware_version()
+                #component_description = component.get_description()
+
+                #module_component_map[module_name][component_name] = component
+                #module_component_map[module_name][component_name] = (
+                #    component_firmware_version,
+                #    component_description
+                #)
+
+        return module_component_map
+
+    def is_modular_chassis(self):
+        return len(self.module_component_map) > 0
+
+    #def is_chassis_has_components(self):
+    #    return len(self.chassis_component_map) > 0
+
+    def is_chassis_has_components(self):
+        #print("debug")
+        #print(self.__chassis.get_num_components())
+        return self.__chassis.get_num_components() > 0
+
+    platform = property(fget=__get_platform)
+    chassis = property(fget=__get_chassis)
+
+
+
+
+class PlatformComponentsParser(object):
+    """
+    PlatformComponentsParser
+    """
+    PLATFORM_COMPONENTS_FILE = "platform_components.json"
+    PLATFORM_COMPONENTS_PATH_TEMPLATE = "{}/usr/share/sonic/device/{}/{}"
+
+    CHASSIS_KEY = "chassis"
+    MODULE_KEY = "module"
+    COMPONENT_KEY = "component"
+    FIRMWARE_KEY = "firmware"
+    VERSION_KEY = "version"
+    INFO_KEY = "info"
+
+    UTF8_ENCODING = "utf-8"
+
+    def __init__(self, is_modular_chassis):
+        self.__is_modular_chassis = is_modular_chassis
+        self.__chassis_component_map = OrderedDict()
+        self.__module_component_map = OrderedDict()
+
+    def __get_platform_type(self):
+        machine_info = sonic_device_util.get_machine_info()
+
+        return sonic_device_util.get_platform_info(machine_info)
+
+    def __get_platform_components_path(self, root_path):
+        platform_type = self.__get_platform_type()
+
+        return self.PLATFORM_COMPONENTS_PATH_TEMPLATE.format(
+            root_path,
+            platform_type,
+            self.PLATFORM_COMPONENTS_FILE
+        )
+
+    def __is_str(self, obj):
+        return isinstance(obj, unicode) or isinstance(obj, str)
+
+    def __is_dict(self, obj):
+        return isinstance(obj, dict)
+
+    def __parser_fail(self, msg):
+        raise RuntimeError("Failed to parse \"{}\": {}".format(self.PLATFORM_COMPONENTS_FILE, msg))
+
+    def __parser_platform_fail(self, msg):
+        self.__parser_fail("invalid platform schema: {}".format(msg))
+
+    def __parser_chassis_fail(self, msg):
+        self.__parser_fail("invalid chassis schema: {}".format(msg))
+
+    def __parser_module_fail(self, msg):
+        self.__parser_fail("invalid module schema: {}".format(msg))
+
+    def __parser_component_fail(self, msg):
+        self.__parser_fail("invalid component schema: {}".format(msg))
+
+    def __parse_component_section(self, section, component, is_module_component=False):
```

Comment:
@stephenxs done

Reply:
@stephenxs that's a good point for optimization, but it's not trivial since we need to provide different objects initialization and show different errors.